beaker/src/stamp_manager.rs

```diff
@@ -0,0 +1,284 @@
+
+use crate::cache_common::{calculate_md5, get_cache_base_dir};
+use crate::config::{CutoutConfig, DetectionConfig};
+use beaker_stamp::{write_cfg_stamp, Stamp};
+
+/// Generate a deterministic configuration hash for detection using proc macro
+///
+/// This function now uses the Stamp trait implementation generated by the derive macro.
+/// The derive macro automatically includes all fields marked with #[stamp] in DetectionConfig.
+pub fn generate_detection_config_hash(config: &DetectionConfig) -> String {
+    let full_hash = config.stamp_hash();
+    // Extract just the hex part (remove "sha256:" prefix) and truncate to 16 chars for consistency
+    full_hash.trim_start_matches("sha256:")[..16].to_string()
+}
+
+/// Generate a deterministic configuration hash for cutout processing using proc macro
+///
+/// This function now uses the Stamp trait implementation generated by the derive macro.
+/// The derive macro automatically includes all fields marked with #[stamp] in CutoutConfig.
+pub fn generate_cutout_config_hash(config: &CutoutConfig) -> String {
+    let full_hash = config.stamp_hash();
+    // Extract just the hex part (remove "sha256:" prefix) and truncate to 16 chars for consistency
+    full_hash.trim_start_matches("sha256:")[..16].to_string()
+}
+
+/// Generate tool version hash
+pub fn generate_tool_hash() -> String {
+    let mut hasher = Sha256::new();
+    hasher.update(format!("beaker:{}", env!("CARGO_PKG_VERSION")));
+    format!("{:x}", hasher.finalize())[..16].to_string()
+}
+
+/// Generate model hash from file path
+pub fn generate_model_hash_from_path(model_path: &Path) -> Result<String> {
+    if model_path.exists() {
+        calculate_md5(model_path).map(|hash| hash[..16].to_string())
+    } else {
+        Err(anyhow!("Model file not found: {}", model_path.display()))
+    }
+}
+
+/// Get the beaker stamps cache directory
+pub fn get_stamps_cache_dir() -> Result<PathBuf> {
+    let cache_dir = get_cache_base_dir()?;
+    Ok(cache_dir.join("beaker").join("stamps"))
+}
+
+/// Create stamp file if content has changed
+pub fn create_or_update_stamp(stamp_type: &str, hash: &str, content: &str) -> Result<PathBuf> {
+    let stamps_dir = get_stamps_cache_dir()?;
+    fs::create_dir_all(&stamps_dir)?;
+
+    let stamp_filename = format!("{stamp_type}-{hash}.stamp");
+    let stamp_path = stamps_dir.join(stamp_filename);
+
+    // Check if stamp already exists with same content
+    if stamp_path.exists() {
+        if let Ok(existing_content) = fs::read_to_string(&stamp_path) {
+            if existing_content == content {
+                // Content is identical, preserve mtime
+                return Ok(stamp_path);
+            }
+        }
+    }
+
+    // Write new content atomically
+    let temp_path = stamp_path.with_extension("tmp");
+    fs::write(&temp_path, content)?;
+    fs::rename(temp_path, &stamp_path)?;
+
+    Ok(stamp_path)
+}
+
+/// Generate all stamps for a detection run using beaker-stamp
```

Comment:
can we make this generic? so don't need 1 function per model, instead 1 function overall